src/Aspire.Hosting/AspireStore.cs

```diff
@@ -0,0 +1,234 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Reflection;
+using System.Security.Cryptography;
+
+namespace Aspire.Hosting;
+
+/// <summary>
+/// Represents a store for managing files in the Aspire hosting environment that can be reused across runs.
+/// </summary>
+/// <remarks>
+/// The store is created in the ./obj folder of the Application Host.
+/// If the ASPIRE_STORE_DIR environment variable is set this will be used instead.
+///
+/// The store is specific to a <see cref="IDistributedApplicationBuilder"/> instance such that each application can't
+/// conflict with others. A <em>.aspire</em> prefix is also used to ensure that the folder can be delete without impacting
+/// unrelated files.
+/// </remarks>
+public class AspireStore
+{
+    internal const string AspireStorePathKeyName = "Aspire:Store:Path";
+
+    private readonly string _basePath;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AspireStore"/> class with the specified base path.
+    /// </summary>
+    /// <param name="basePath">The base path for the store.</param>
+    private AspireStore(string basePath)
+    {
+        ArgumentNullException.ThrowIfNull(basePath);
+
+        _basePath = basePath;
+        EnsureDirectory();
+    }
+
+    /// <summary>
+    /// Gets the base path of this store.
+    /// </summary>
+    public string BasePath => _basePath;
+
+    /// <summary>
+    /// Creates a new instance of <see cref="AspireStore"/> using the provided <paramref name="builder"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <returns>A new instance of <see cref="AspireStore"/>.</returns>
+    public static AspireStore Create(IDistributedApplicationBuilder builder)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        var assemblyMetadata = builder.AppHostAssembly?.GetCustomAttributes<AssemblyMetadataAttribute>();
```

Comment:
We don't need to do this if `AspireStorePathKeyName` is set. So we can put it inside an `if`.